examples/QKD/e91_combined.py

```diff
@@ -0,0 +1,294 @@
+    print("Bob Key Decoy 3_0 a1b3: {}".format(bob_key_decoy_a1b3))
+    print("Bob Key Decoy 4_2 a3b1: {}".format(bob_key_decoy_a3b1))
+    print("Bob Key Decoy 3_2 a3b3: {}".format(bob_key_decoy_a3b3))
+    bob_key_decoy.append(bob_key_decoy_a1b1)
+    bob_key_decoy.append(bob_key_decoy_a1b3)
+    bob_key_decoy.append(bob_key_decoy_a3b1)
+    bob_key_decoy.append(bob_key_decoy_a3b3)
+
+    print("Bob Key Raw: {}".format(bob_key_raw))
+    print("Bob Key Raw 4 a2b1: {}".format(bob_key_raw_a2b1))
+    print("Bob Key Raw 2 a3b2: {}".format(bob_key_raw_a3b2))
+    return bob_key_decoy, bob_key_raw_a2b1, bob_key_raw_a3b2, bob_key_raw
+
+
+def alice(host, receiver, alice_basis):
+    alice_measured_bits = ""
+    # For Qubit and Basis
+    for basis in alice_basis:
+        q1, q2 = entangle(host)
+        ack_arrived = host.send_qubit(receiver, q2, await_ack=False)
+        if ack_arrived:
+            if basis == '0':    # 0
+                q1.ry(0.0)
+                alice_measured_bits += str(q1.measure())
+            if basis == '4':    # pi/4
+                q1.ry(0.785398)
+                alice_measured_bits += str(q1.measure())
+            if basis == '2':    # pi/2
+                q1.ry(1.5708)
+                alice_measured_bits += str(q1.measure())
+    print("Alice's measured bits: {}".format(alice_measured_bits))
+
+    # Sending Basis to Bob
+    ack_basis_alice = host.send_classical(receiver, alice_basis, await_ack=True)
+    if ack_basis_alice is not None:
+        print("{}'s basis string successfully sent".format(host.host_id))
+    # Receiving Basis from Bob
+    basis_from_bob = host.get_next_classical(receiver, wait=WAIT_TIME)
+    if basis_from_bob is not None:
+        print("{}'s basis string got successfully by {}".format(receiver, host.host_id))
+
+    # For Key
+    alice_key_decoy, alice_key_raw_a2b1, alice_key_raw_a3b2, alice_key_raw = alice_key_string(alice_measured_bits, alice_basis, basis_from_bob.content)
+    print("Alice Key Decoy: {}".format(alice_key_decoy))
+
+    # For Sending Decoy Key
+    alice_brd_ack = host.send_classical(receiver, alice_key_decoy, await_ack=True)
+    if alice_brd_ack is not None:
+        print("{}'s key successfully sent to {}".format(host.host_id, receiver))
+    bob_key_decoy = host.get_next_classical(receiver, wait=WAIT_TIME)
+    if bob_key_decoy is not None:
+        print("{}'s got successfully by {}".format(receiver, host.host_id))
+        print("Bob Key Decoy got by Alice: {}".format(bob_key_decoy.content))
+
+    # For Testing CHSH Inequality
+    result = bell_state(alice_key_decoy, bob_key_decoy.content)
+    print("Result Alice: {}".format(result))
+    if 2.0 <= abs(result) < 4:  # nearly equivalent to 2.82
+        print("|S| value: {}".format(abs(result)))
+    else:
+        print("Alice and Bob discard the bits and run again")
+
+
+def eve_sniffing_quantum(sender, receiver, qubit):
+    qubit.measure(non_destructive=True)
```

Comment:
Maybe something interesting is to measure only some of the qubits, like with some probability `p`, Eve measures a qubit, then you can see which values of `p` the protocol accepts the key.

Reply:
Ok. Let me try. I will let you know if I can do it or else you can merge it.

Reply:
Seems like I cannot do it. You can merge it. I won't add it.